### ISSUE TITLE
WinPhone8/UWP/8.1 - wams_state cookie missing on MSA redirect

### ISSUE THREAD

**Issue by aziel** [Azure/azure-mobile-services#891]
Some customers report that wams_state cookie is missing on the redirect from MSA using our client SDK.

[A customer posted on March 21 experiencing this with UWP](https://social.msdn.microsoft.com/Forums/en-US/6bbdb774-29ba-4205-ba97-8d22de4c4795/login-failed-error-redirect-from-microsoftaccount-does-not-contain-the-required-wamsstate-cookie?forum=azuremobile&prof=required). 

The latest customer that I am working with reported some of his customers were unable to login due to this issue ~ 4/4/2016. One customer worked with me to give me an account login and password that consistently reproes the issue on WP8 client (which uses a browser).  

The scenario that causes the error is as follows:
1. User initiates login with Microsoft Account.
    - client makes call to site/login/microsoftaccount
2. User sees dialog box and enters username and password.  User clicks Sign-in button.
3. Nothing appears to change happens on the client.  It appears that sign-in did not happen to the client.  
    - In actuality, (verified with Fiddler), the client continued the login process, and received login/done with the auth token.  
4. Since nothing happened in Step 3.  Customer clicks sign-in again.
    - In fiddler, it appears the client continued the login process, but the final redirect does not contain the wams_state cookie.
5. Client does not find wams_state cookie in the 302, and throws and exception.

This does not appear to happen if you hit the browser directly (Chrome or IE), and go through the /login/microsoftaccount process.

**Comment by aziel** [comment]
Workaround for this is to use Live SDK to get the authenticationToken from Microsoft Account, and trade it for a Zumo token using the REST API.
An example using C# and Live SDK:

```c#
 MobileServiceUser user = null;
            if (provider == MobileServiceAuthenticationProvider.MicrosoftAccount)
            {
                // Use Live SDK to get the auth token
                LiveLoginResult result = null;
                try
                {
                    // Get token from Live
                    LiveAuthClient authClient = new LiveAuthClient("YOURLIVEAPPCLIENTID");
                    result = await authClient.LoginAsync(new List<string> { "wl.signin" });

                    if (result != null && result.Status == LiveConnectSessionStatus.Connected)
                    {
                        LiveConnectSession session = result.Session;
                        string token = session.AuthenticationToken;

                        // Client directed login flow
                        // See https://msdn.microsoft.com/en-us/library/azure/jj710106.aspx
                        // Trade live authenticationToken for Zumo token
                        using (HttpClient client = new HttpClient())
                        {
                            string clientDirectedLoginUri = "https://YOURMOBILESERVICENAME.azure-mobile.net/login/microsoftaccount";
                            using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, clientDirectedLoginUri))
                            {
                                request.Content = new StringContent("{\"authenticationToken\":\"" + token + "\"}", Encoding.UTF8, "application/json");
                                using (HttpResponseMessage response = await client.SendAsync(request))
                                {
                                    if (response.IsSuccessStatusCode)
                                    {
                                        // Set the MobileServiceUser
                                        string responseContent = await response.Content.ReadAsStringAsync();
                                        JObject jobj = JObject.Parse(responseContent);
                                        string parsedId = (string)jobj.SelectToken("user.userId");
                                        MobileServiceUser parsedUser = new MobileServiceUser(parsedId);
                                        parsedUser.MobileServiceAuthenticationToken = (string)jobj.SelectToken("authenticationToken");

                                        user = parsedUser;
                                    }
                                    else
                                    {
                                        throw new InvalidOperationException("Live SDK login failed");
                                    }
                                } // dispose response
                            } // dispose request
                        } // dispose client
                    }
                }
                catch 
                {
                    // Catches exceptions from Live login or Zumo login
                    result = null;
                }
            }
            else
            {
                // All other providers will use MobileServiceClient
                try
                {
                    user = await EntitiesTools.MobileService.LoginAsync(provider);
                }
                catch (Exception ex)
                {
                    user = null;
                }
            }
```

**Comment by knaopel** [comment]
Hello @aziel ,

This solution is __not__ working for me. I added the Live SDK nugget package and added the code to the AuthenticateAsync method and it throws a System.ArgumentException error. I am trying to sign in from a cached login information and if that fails directs the user to login. Below is the method I am using to call AuthenticateAsync in the MainPage constructor.

            Loaded += async (sender, args) =>
            {
                await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, async () =>
                {
                    await AuthenticateAsync(Microsoft.WindowsAzure.MobileServices.MobileServiceAuthenticationProvider.MicrosoftAccount);
                    await RefreshItems();
                });
            };

Also included is the AuthenticateAsync method below:

        private async Task AuthenticateAsync(MobileServiceAuthenticationProvider Provider)
        {
            string message;

            PasswordVault vault = new PasswordVault();
            PasswordCredential credential = null;

            InvalidOperationException exception = null;

            while (credential == null)
            {
                try
                {
                    // try getting credential from vault.
                    credential = vault.FindAllByResource(Provider.ToString()).FirstOrDefault();
                }
                catch (Exception)
                {
                    // when there is no matching resource it throws an error which we ignore.
                }
                if (credential != null)
                {
                    user = new MobileServiceUser(credential.UserName);
                    credential.RetrievePassword();
                    user.MobileServiceAuthenticationToken = credential.Password;

                    // set the user from the stored credentials
                    App.MobileService.CurrentUser = user;

                    try
                    {
                        // try to return an item to see if the cached credential has expired.
                        await App.MobileService.GetTable<TimeCardItem>().Take(1).ToListAsync();
                    }
                    catch (MobileServiceInvalidOperationException ex)
                    {
                        if (ex.Response.StatusCode == System.Net.HttpStatusCode.Unauthorized)
                        {
                            // remove the credential with the expired token.
                            vault.Remove(credential);
                            credential = null;
                            continue;
                        }   
                    }
                }
                else
                {
                    try
                    {
                        // inform the user that they must log in
                        var fyiDialog = new MessageDialog("Please (re)login with your Microsoft Account credentials.", "Login Required");
                        fyiDialog.Commands.Add(new UICommand("OK"));
                        await fyiDialog.ShowAsync();

                        // use Live SDK to get auth token
                        LiveLoginResult result = null;
                        //try
                        //{
                            // Get token from live
                            LiveAuthClient authClient = new LiveAuthClient("[clientIdHere]");
                            result = await authClient.LoginAsync(new List<string> { "wl.signin" });
                            if(result !=null && result.Status == LiveConnectSessionStatus.Connected)
                            {
                                LiveConnectSession session = result.Session;
                                string token = session.AuthenticationToken;

                                //clinet directed login flow
                                using(HttpClient client = new HttpClient())
                                {
                                    string clientDirectedLoginUri = "https://[zumoNameHere].azure-mobile.net/login/microsoftaccount";
                                    using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, clientDirectedLoginUri))
                                    {
                                        request.Content = new StringContent("{\"authenticationToken\": \"" + token + "\"}");
                                        using(HttpResponseMessage response = await client.SendAsync(request))
                                        {
                                            if (response.IsSuccessStatusCode)
                                            {
                                                // set the mobile services user
                                                string responseContent = await response.Content.ReadAsStringAsync();
                                                JObject jObj = JObject.Parse(responseContent);
                                                string parseId = (string)jObj.SelectToken("user.userId");
                                                MobileServiceUser parsedUser = new MobileServiceUser(parseId);
                                                parsedUser.MobileServiceAuthenticationToken = (string)jObj.SelectToken("authenticationToken");

                                                user = parsedUser;
                                            }
                                            else
                                            {
                                                throw new InvalidOperationException("Live SDK Login failed");
                                            }
                                        } // dispose HttpResponseMessage
                                    } // dispose HttpRequestMessage
                                } // dispose HttpClient
                            }
                        //}
                        //catch(InvalidOperationException e)
                        //{
                        //    exception = e;
                        //    result = null;
                        //}

                        // sign-in using the Identity Provider
                        //user = await App.MobileService.LoginAsync(Provider);

                        // create the user and store the credential
                        credential = new PasswordCredential(Provider.ToString()
                            , user.UserId
                            , user.MobileServiceAuthenticationToken);
                        vault.Add(credential);
                        //message = string.Format("You are logged in = {0}", user.UserId);
                    }
                    catch (InvalidOperationException e)
                    {
                        exception = e;
                    }
                    if(exception != null)
                    {
                        message = "You must log in. login required.";
                        var dialog = new MessageDialog(message);
                        dialog.Commands.Add(new UICommand("OK"));
                        await dialog.ShowAsync();
                    }
                }
            }
        }

Any insight on how to get this working again - my app is __not__ working!
Thanks!
Kurt

**Comment by dhei** [comment]
@knaopel, do you know where is System.ArgumentException thrown? Is it `result = await authClient.LoginAsync(new List<string> { "wl.signin" });` or can you provide the stacktrace? Are you checking null parameter before invoking authclient login?

**Comment by knaopel** [comment]
@dhei the exception was being thrown from `LiveAuthClient authClient = new LiveAuthClient("[clientIdHere]");`. I have removed the clientId as you suggested, but now on the line `string token = session.AuthenticationToken;` the AuthenticationToken is null and therefore the call to login to AzureMobileServices comes back unauthenticated. Any Ideas?

>

**Comment by aziel** [comment]
@knaopel 
Hi Kurt, Sorry, I was on vacation when you initially asked the question.

Di pointed out an issue with the documentation.
My workaround is specifically for WP8.

Note that Live SDK LiveAuthClient constructor is different for WP8: [WP8 LiveAuthClient](https://github.com/liveservices/LiveSDK-for-Windows/blob/ff7759a3e650506b98fdd7f2980cde03e559ca84/src/WP8/Source/Public/LiveAuthClient.cs#L48)

[WinStore LiveAuthClient](https://github.com/liveservices/LiveSDK-for-Windows/blob/ff7759a3e650506b98fdd7f2980cde03e559ca84/src/WinStore/Source/Public/LiveAuthClient.cs#L59)

The meaning of the string parameter is different.

Which platform are you developing for?

**Comment by knaopel** [comment]
It is a 8.1 universal App but right now my concern in Windows Phone 8.1 (Windows 8.1 App has not been released yet, WP8.1 App is in the store - and **not** working) :-(
Thanks for any help.

**Comment by aziel** [comment]
@knaopel 
Can you show me your packages.config?  I'm curious what version of LiveSDK you're trying to use.

If you could email me privately at azepil at microsoft dot com, we can try to resolve this more rapidly.

**Comment by knaopel** [comment]
@aziel my packages.config is as follows:
`<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="LiveSDK" version="5.6.3" targetFramework="wpa81" />
  <package id="Microsoft.Bcl" version="1.1.9" targetFramework="wpa81" />
  <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="wpa81" />
  <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="wpa81" />
  <package id="Newtonsoft.Json" version="6.0.4" targetFramework="wpa81" />
  <package id="WindowsAzure.MobileServices" version="1.3.2" targetFramework="wpa81" />
</packages>`
As you can see the LiveSDK is v5.6.3. Thanks!

**Comment by aziel** [comment]
Working with Kurt offline, we found that he is using WinPhone 8.1 client.
The LiveSDK constructor is different for 8.1 and 8.0.
8.0 uses 
LiveAuthClient authClient = new LiveAuthClient("YOURLIVEAPPCLIENTID");
which is the same as in my code.

8.1 uses
LiveAuthClient authClient = new LiveAuthClient("YourLiveAppRedirectUri");
which is something of the form https://myapp.azure-mobile.net/signin-microsoftaccount

**Comment by mkm-dz** [comment]
What about UWP, I can repro it on UWP with the latest WindowsAzure.MobileServices (1.3.2) work around does not work for me because LiveSDK should not be used on UWP. (https://github.com/liveservices/LiveSDK-for-Windows/issues/58)

**Comment by aziel** [comment]
Try OneDrive SDK for Microsoft Account auth on UWP.  The thread you linked has a reference to OneDrive SDK.

You should be able to exchange the access token for a Zumo token in the same manner.

/iPhone

**Comment by mkm-dz** [comment]
Thanks for the quick follow up, I'll try it and let you know the outcome. Just as a suggestion, this thread should not be closed as the issue is still active, we have a workaround but it is still active.

**Comment by mkm-dz** [comment]
So I used Onedrive SDK via `OneDriveClientExtensions.GetClientUsingWebAuthenticationBroker`
and was able to get the token, but when I reach 

`using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, clientDirectedLoginUri))
                                {
                                    request.Content = new StringContent("{\"authenticationToken\": \"" + token + "\"}");
                                    using(HttpResponseMessage response = await client.SendAsync(request))
                                    {
`
I get an 401, unauthorized. Any idea?

**Comment by adrianhall** [comment]
We have no plans to support UWP on Azure Mobile Services.  You can upgrade your site to Azure Mobile Apps to get official UWP Support.

**Comment by mkm-dz** [comment]
Hi adrianhall, now I'm completely lost :( I got no idea what should I use, I have a UWP which was working and now is broken. I follow this tutorial : https://azure.microsoft.com/en-us/documentation/articles/app-service-mobile-windows-store-dotnet-get-started-users/

which seems to be the official one and they point me ot use this. There are WindowsAzure.MobileServices nuget, Microsoft.Azure.Mobile.Client nuget and ADAL nuget and the last comment points me to use Onedrive SDK! Is there an updated version on how to migrate my (previously working) UWP & Azure Mobile Services to Azure Mobile Apps?

**Comment by adrianhall** [comment]
@mkm-dz - you are already using Azure Mobile Apps.  Mobile Services is accessed through the classic portal (manage.windowsazure.com) and is in the process of being deprecated.  Mobile Apps (and App Service) are the successor to that service.

**Comment by mkm-dz** [comment]
thank you  for the follow up @adrianhall, I migrated my service but for some reason following tutorial does not work with a migrated service I got the same exception. I created a new service from scratch, migrated my backend and change my rest api to use the new zumo v2 and now it works (same code, just with a different service) my guess is migration broke something. Not the best dev experience but at least is working now.
